File: express-backend/src/models/Message.ts
import mongoose, { Schema, type Document, type Model } from 'mongoose';

export interface MessageDocument extends Document {
  conversation_id: mongoose.Types.ObjectId;
  sender_type: string;
  sender_user_id?: mongoose.Types.ObjectId;
  body: string;
  attachment_ids?: mongoose.Types.ObjectId[];
  created_at: Date;
}

const MessageSchema = new Schema<MessageDocument>(
  {
    conversation_id: {
      type: Schema.Types.ObjectId,
      ref: 'Conversation',
      required: true,
      index: true,
    },
    sender_type: { type: String, required: true },
    sender_user_id: { type: Schema.Types.ObjectId, ref: 'User' },
    body: { type: String, required: true },
    attachment_ids: [{ type: Schema.Types.ObjectId, ref: 'Blob' }],
    created_at: { type: Date, default: () => new Date() },
  },
  {
    collection: 'messages',
  },
);

export const Message: Model<MessageDocument> =
  mongoose.models.Message || mongoose.model<MessageDocument>('Message', MessageSchema);

File: backend/app/schemas/message.py
"""Chat message document schema."""
from datetime import datetime, timezone
from typing import List, Optional

from pydantic import BaseModel, Field

from .enums import SenderType
from .objectid import PyObjectId


def _utc_now() -> datetime:
    return datetime.now(timezone.utc)


class MessageDoc(BaseModel):
    """Mongo document shape for messages collection. sender_user_id only for provider messages."""

    id: Optional[PyObjectId] = Field(None, alias="_id")
    conversation_id: PyObjectId
    sender_type: SenderType
    sender_user_id: Optional[PyObjectId] = None
    body: str
    attachment_ids: Optional[List[PyObjectId]] = None
    created_at: datetime = Field(default_factory=_utc_now)

    model_config = {"populate_by_name": True, "arbitrary_types_allowed": True}
